File: packages/react-native-nitro-web-image/ios/HybridWebImageLoader.swift
//
//  HybridWebImageLoader.swift
//  NitroWebImage
//
//  Created by Marc Rousavy on 28.07.25.
//

import NitroModules
import SDWebImage
import NitroImage

fileprivate class HybridImage: HybridImageSpec, NativeImage {
  let uiImage: UIImage
  init(uiImage: UIImage) {
    self.uiImage = uiImage
    super.init()
  }
}

class HybridWebImageLoader: HybridImageLoaderSpec {
  private let url: URL
  private let options: AsyncImageLoadOptions?

  init(url: URL, options: AsyncImageLoadOptions?) {
    self.url = url
    self.options = options
    super.init()
  }

  func loadImage() throws -> Promise<(any HybridImageSpec)> {
    return Self.loadImage(url: self.url, options: self.options)
  }

  func requestImage(forView view: (any HybridNitroImageViewSpec)) throws {
    guard let imageView = (view as? NativeImageView)?.imageView else { throw RuntimeError.error(withMessage: "Invalid view type!") }

    var webImageOptions = options?.toSDWebImageOptions() ?? []
    switch view.priority {
    case 0: webImageOptions.insert(.lowPriority)
    case 2: webImageOptions.insert(.highPriority)
    default: break
    }
    let webImageContext = options?.toSDWebImageContext()
    imageView.sd_setImage(with: url,
                          placeholderImage: imageView.image,
                          options: webImageOptions,
                          context: webImageContext)
  }

  func dropImage(forView view: (any HybridNitroImageViewSpec)) throws {
    // TODO: Do we need to reset the image here or not?
  }

  public static func loadImage(url: URL, options: AsyncImageLoadOptions?) -> Promise<any HybridImageSpec> {
    return Promise.async {
      let uiImage = try await SDWebImageManager.shared.loadImage(with: url,
                                                                 options: options)
      return HybridImage(uiImage: uiImage)
    }
  }
}

File: packages/react-native-nitro-web-image/android/src/main/java/com/margelo/nitro/web/image/HybridWebImageLoader.kt
package com.margelo.nitro.web.image

import android.content.Context
import android.widget.ImageView
import coil3.ImageLoader
import coil3.load
import com.margelo.nitro.core.Promise
import com.margelo.nitro.image.HybridImageSpec
import com.margelo.nitro.image.HybridImageLoaderSpec
import com.margelo.nitro.image.HybridNitroImageViewSpec
import com.margelo.nitro.web.image.cache.CACHE_PRIORITY_EXTRA
import com.margelo.nitro.web.image.extensions.applyOptions
import com.margelo.nitro.web.image.extensions.loadImageAsync

class HybridWebImageLoader(private val imageLoader: ImageLoader,
                           private val url: String,
                           private val options: AsyncImageLoadOptions?,
                           private val context: Context) : HybridImageLoaderSpec() {
    override fun loadImage(): Promise<HybridImageSpec> {
        return imageLoader.loadImageAsync(url, options, context)
    }

    override fun requestImage(forView: HybridNitroImageViewSpec) {
        val imageView = forView.view as? ImageView ?: return
        val cachePriority = forView.cachePriority?.toInt() ?: 1

        imageView.load(url, imageLoader) {
            this.applyOptions(options)
            memoryCacheKeyExtra(CACHE_PRIORITY_EXTRA, cachePriority.toString())
        }
    }

    override fun dropImage(forView: HybridNitroImageViewSpec) {
        // Coil automatically handles recycling here - I _think_.
    }
}
